lib/modules/filteReddit.js

```diff
@@ -732,338 +947,817 @@ module.go = () => {
+					isCurrentSubreddit('all') || currentDomain() || isCurrentMultireddit('me/f/all')) ? false : null,
+				evaluate(thing) { return filtersMatchString(this.key, thing.getSubreddit() || currentSubreddit(), null, true); },
+				clearCache() { getStringMatchFilters.cache.delete(this.key); },
+			}, {
+				key: 'flair',
+				evaluate(thing) { return filtersMatchString(this.key, thing.getPostFlairText(), thing.getSubreddit()); },
+				clearCache() { getStringMatchFilters.cache.delete(this.key); },
+			}, {
+				key: 'customFilters',
+				evaluate: executeCustomFilters,
+			},
+		];
+
+		for (const options of externalFilters) {
+			filterline.addFilter(new ExternalFilter({
+				state: false, // Hide all that matches
+				settingsHtml: SettingsNavigation.makeUrlHashLink(module.moduleID, options.key, ' ', 'gearIcon'),
+				...options,
+			}));
 		}
-	);
+	}
 
-	const getSubreddit = val => (
-		val ||
-		SelectedEntry.selectedThing() && SelectedEntry.selectedThing().getSubreddit() ||
-		currentSubreddit() ||
-		''
-	);
+	function addPostCaseFilters() {
+		for (const [key, options] of Object.entries(postCases)) {
+			if (options.disabled) continue;
 
-	CommandLine.registerCommand(/^f(?:ilter)?/, 'f[ilter] [subreddit] - toggle subreddit filter',
-		(cmd, val) => {
-			const subreddit = getSubreddit(val);
-			return `toggle subreddit filter for: ${subreddit}`;
-		},
-		(cmd, val) => {
-			const subreddit = getSubreddit(val);
-			if (!subreddit) return 'no subreddit specified or post selected';
-			toggleFilter(subreddit);
+			options.key = key;
+
+			filterline.addChoice(options);
+			if (options.alwaysShow) filterline.createFilterFromKey(key, { add: true, id: key });
 		}
-	);
+	}
+
+	const completeFilterline = _.once(() => {
+		addPostCaseFilters();
+		if (filterlineState) filterline.restoreState(filterlineState);
+		filterline.createElement();
+	});
+
+	const filterlineTab = CreateElement.tabMenuItem({
+		text: '∀',
+		title: 'Toggle filterline',
+		className: 'res-toggle-filterline-visibility',
+	});
+
+	let active;
+
+	filterlineTab.addEventListener('change', e => {
+		active = e.detail;
+
+		if (active) {
+			completeFilterline();
+			filterlineTab.removeAttribute('aftercontent');
+		} else {
+			const anyCustomFiltersEnabled = filterline.getActiveFilters().find(v => !(v instanceof ExternalFilter));
+			filterlineTab.setAttribute('aftercontent', anyCustomFiltersEnabled ? ' (active)' : '');
+		}
+
+		Storage.patch(filterlineKey, { active });
+		BodyClasses.toggle(active, 'res-filteReddit-show-filterline');
+	});
+
+	addExternalFilters();
+
+	// When using the command line, show the filterline
+	filterline.registerCommand(() => { if (!active) filterlineTab.click(); });
+
+	if (initialFilterlineVisibility) {
+		filterlineTab.click();
+		BodyClasses.toggle(false, 'res-filteReddit-filterline-pad-until-ready');
+	}
+
+	return filterline;
+}
+
+class Filter {
```

Comment:
I think it would make sense to move these classes, the big object definitions at the top (`browseContexts`, etc.) to their own files in `lib/modules/filteReddit` and import them here; should make things easier to manage